pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,221 @@
+        # logs publish status
+        for task in tasks:
+            LOG.info(
+                "Publishing %s %s",
+                repo.id,
+                "successful" if task.succeeded else "unsuccessful",
+            )
+
+    @step("Flush UD cache")
+    def flush_ud(self, repos):
+        client = self.udcache_client
+        if not client:
+            LOG.info("UD cache flush is not enabled.")
+            return []
+
+        out = []
+        for repo in repos:
+            out.append(client.flush_repo(repo.id))
+            if repo.eng_product_id:
+                out.append(client.flush_product(repo.eng_product_id))
+
+        return out
+
+    @step("Flush CDN cache")
+    def flush_cdn(self, repos):
+        if not self.fastpurge_client:
+            LOG.info("CDN cache flush is not enabled.")
+            return []
+
+        def purge_repo(repo):
+            to_flush = []
+            for url in repo.mutable_urls:
+                flush_url = os.path.join(
+                    self.fastpurge_root_url, repo.relative_url, url
+                )
+                to_flush.append(flush_url)
+
+            LOG.debug("Flush: %s", to_flush)
+            flush = self.fastpurge_client.purge_by_url(to_flush)
+            return f_map(flush, lambda _: repo)
+
+        return [purge_repo(r) for r in repos if r.relative_url]
+
+    @step("Get repos")
+    def get_repos(self):
+        repos = self.args.repos
+        found_repo_ids = []
+        out = []
+
+        # apply the filters and get repo_ids
+        repo_ids = self._filter_repos(repos)
+
+        # get repo objects for the repo_ids
+        searched_repos = self.pulp_client.search_repository(Criteria.with_id(repo_ids))
+        for repo in searched_repos:
+            out.append(repo)
+            found_repo_ids.append(repo.id)
+
+        # Bail out if user requested repos which don't exist
+        # or there are no repos returned to publish
+        missing = set(repo_ids) - set(found_repo_ids)
+
+        missing = sorted(list(missing))
+        if missing:
+            self.fail("Requested repo(s) don't exist: %s", ", ".join(missing))
+
+        if not out:
+            self.fail("No repo(s) found to publish")
+
+        return out
+
+    def fail(self, *args, **kwargs):
+        LOG.error(*args, **kwargs)
+        sys.exit(30)
+
+    def _filter_repos(self, repos):
+
+        repos = set(repos) if repos else set()
+
+        if self.args.published_before or self.args.url_regex:
+            repo_dist_f = self._filtered_repo_distributors()
+            filtered_repos = [repo_dist.repo_id for repo_dist in repo_dist_f]
+
+            return list(repos.intersection(filtered_repos)) if repos else filtered_repos
+
+        return list(repos)
+
+    def _filtered_repo_distributors(self):
+        published_before = self.args.published_before
+        url_regex = self.args.url_regex
+
+        # define the criteria on available filters
+        if published_before and url_regex:
```

Comment:
Arranging the code this way looks a little odd and not scalable - how will this code look if we add more distributor search clauses, like published_after?  Will there be one "if" case for every possible combination of arguments?

Would it rather be better to write the criteria for each branch just once, e.g.

```
crit = Criteria.true()
if published_before:
  crit = Criteria.and_([crit, Criteria.with_field("last_publish", Matcher.less_than(published_before)])
if url_regex:
  crit = Criteria.and_([crit, Criteria.with_field("relative_url", Matcher.regex(url_regex)])
```

Avoids having to write out the criteria for each argument more than once, and the number of ``if`` clauses is ``O(N)`` in terms of the arguments, instead of ``O(N^2)``.

Reply:
agree. have made the changes as suggested.